### ISSUE TITLE
JSX Improvements

### ISSUE THREAD

**Issue by acidicX** [peerigon/eslint-config-peerigon#5]
> `'react/no-multi-comp': [2, { ignoreStateless: true }],`

I think it is OK to define more than one functional stateless component in one file, but not more than one stateful component per file.

> `'react/jsx-no-duplicate-props': [2, { ignoreCase: true }],`

Ignore case is probably good, otherwise it would lead to bad readability (passing `isFetching` and `isfetching` should be avoided)?

---

This is more or less personal preference:

> `'react/jsx-curly-spacing': [2, 'always', { allowMultiline: true }],`

JSX code is often better readable with spacing (e.g. string concat inside curlies)

> `'react/jsx-closing-bracket-location': [2, 'line-aligned'],`

I think it looks nicer, more consistent with JavaScript indentation.

**Comment by flootr** [comment]
> I think it is OK to define more than one functional stateless component in one file

I do not think we should make this a general rule. Having multiple components in one file brings up a few different problems. One is naming the file. Another one is exporting the components if another component depends on them. In the end things get messy and it's unclear where a specific component is located.

> passing isFetching and isfetching should be avoided

I think so. Where is the difference between these two props? 

> JSX code is often better readable with spacing 

I don't think this adds better readability here - also mostly a personal preference here. Maybe this could lead into confusion width object declaration like `const obj = { one: '1' }`?

> 'react/jsx-closing-bracket-location': [2, 'line-aligned'],

👍

**Comment by acidicX** [comment]
> I do not think we should make this a general rule. Having multiple components in one file brings up a few different problems. One is naming the file. Another one is exporting the components if another component depends on them. In the end things get messy and it's unclear where a specific component is located.

I think it is OK if you don't export the component that is not in the name of the file. The file name should always indicate the main component. For example, I have a small helper component, three lines, and used often within the main component. I am sure that it will never be used outside of this file, so I can spare myself all the files (scss as well) and import/export stuff. Keep in mind though that currently this rule is completely disabled right now, so we could use this setting to tighten the restrictions a bit and see how it goes.

> Maybe this could lead into confusion width object declaration

I don't really think that would lead to confusion. Spaces make it better to read because you separate "real code" visually a bit more from the HTML-like JSX stuff.

**Comment by jhnns** [comment]
I agree with everything.

`react/jsx-curly-spacing` is a personal preference, but we're always using spaces when curly braces occur in the code, like `import { bla } from "blub"` or `{ a: true, b: true }`. So that would be consistent.

Regarding `react/jsx-closing-bracket-location`: `line-aligned` means that it should be written this way:

```html
<Hello
  firstName="John"
  lastName="Smith"
/>;

<Say
  firstName="John"
  lastName="Smith"
>
  Hello
</Say>;

var x = <Hello
  firstName="John"
  lastName="Smith"
/>;

var x = function() {
  return <Say
    firstName="John"
    lastName="Smith"
  >
    Hello
  </Say>;
};
```

Right?

**Comment by acidicX** [comment]
> So that would be consistent.

:+1: 

> Right?

yup!

**Comment by flootr** [comment]
> So that would be consistent.

It wouldn't be inconsistent if we don't declare it as general rule. Only if we use different styles for the same thing. Importing modules an defining propTypes aren't the same. Thus, I'm not happy with this rule as general.

**Comment by jhnns** [comment]
> Importing modules an defining propTypes aren't the same

That's true, they are not the same thing. But I like simple rules. And a simple rules is: "After every { and before every } there is a space". Simple rules will last longer as coding convention than complex rules.

**Comment by jhnns** [comment]
I shipped these rules with 9.0.0. I also added a rule that we should use spaces around curly braces in template strings since this is basically the same situation as jsx.
